zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -270,69 +270,156 @@ impl DbFormatChange {
+                .iter()
+                .find(|(_, upgrade)| **upgrade == NetworkUpgrade::Sapling)
+                .expect("there should be sapling upgrade");
+            let (&orchard_height, _) = NetworkUpgrade::activation_list(network)
+                .iter()
+                .find(|(_, upgrade)| **upgrade == NetworkUpgrade::Nu5)
+                .expect("there should be Nu5 upgrade");
+
+            // Delete duplicates before sapling and orchard heights
+            let (mut prev_sapling_tree, mut prev_orchard_tree) = {
+                let height = Height(1);
                 let mut batch = DiskWriteBatch::new();
-                for _ in 0..1_000 {
-                    if height >= initial_tip_height {
-                        break;
-                    }
-                    let sapling_tree = upgrade_db.sapling_tree_by_height(&height);
-                    let orchard_tree = upgrade_db.orchard_tree_by_height(&height);
 
-                    if prev_sapling_tree == sapling_tree {
-                        batch.delete_sapling_tree(&sapling_cf, &height);
-                    }
+                batch.delete_range_sapling_tree(&upgrade_db, &height, &sapling_height);
+                batch.delete_range_orchard_tree(&upgrade_db, &height, &orchard_height);
+                upgrade_db
+                    .write_batch(batch)
+                    .expect("Deleting note commitment trees should always succeed.");
+
+                warn!(?sapling_height, "Database upgrade is at:");
+
+                (
+                    upgrade_db.sapling_tree_by_height(&Height(0)),
+                    upgrade_db.orchard_tree_by_height(&Height(0)),
+                )
+            };
 
-                    if prev_orchard_tree == orchard_tree {
-                        batch.delete_orchard_tree(&orchard_cf, &height);
+            // Create an unbounded channel for reading note commitment trees
+            let (sapling_tree_tx, sapling_tree_rx) = mpsc::channel();
+
+            // Set up task for reading sapling note commitment trees
+            let db = upgrade_db.clone();
+            let sapling_read_task = std::thread::spawn(move || {
+                for (height, tree) in
+                    db.sapling_tree_by_height_range(sapling_height..initial_tip_height)
+                {
+                    let _ = sapling_tree_tx.send((height, tree));
+                }
+            });
+
+            // Create an unbounded channel for duplicate sapling note commitment tree heights
+            let (unique_sapling_tree_height_tx, unique_sapling_tree_height_rx) = mpsc::channel();
+
+            // Set up task for reading sapling note commitment trees
+            let sapling_compare_task = std::thread::spawn(move || {
+                while let Ok((height, tree)) = sapling_tree_rx.recv() {
+                    let tree = Some(tree);
+                    if prev_sapling_tree != tree {
+                        let _ = unique_sapling_tree_height_tx.send(height);
```

Comment:
Should these sends panic if they fail?

Reply:
Logged warning if one of these sends fail in https://github.com/ZcashFoundation/zebra/pull/7321/commits/3aa19d2084f3b9dc1cbfec4f6c49c21a79c491a3